textattack/search_methods/greedy_word_swap_wir.py

```diff
@@ -69,12 +71,47 @@ def _get_index_order(self, initial_text):
                 delta_ps.append(max_score_change)
 
             index_scores = softmax_saliency_scores * np.array(delta_ps)
+
         elif self.wir_method == "delete":
             leave_one_texts = [
                 initial_text.delete_word_at_index(i) for i in range(len_text)
             ]
             leave_one_results, search_over = self.get_goal_results(leave_one_texts)
             index_scores = np.array([result.score for result in leave_one_results])
+
+        elif self.wir_method == "gradient":
+            victim_model = self.get_model()
+            index_scores = np.zeros(initial_text.num_words)
+            grad_output = victim_model.get_grad(initial_text.tokenizer_input)
+            gradient = grad_output["gradient"]
+            j = 0
+            last_matched = 0
+            for i, word in enumerate(initial_text.words):
+                word = initial_text.words[i].lower()
+                matched_tokens = []
+                a = []
+                while j < len(grad_output["tokens"]) and len(word) > 0:
+                    token = grad_output["tokens"][j].lower()
+                    # remove "##" if it's a subword
+                    token = token.replace("##", "")
+                    idx = word.find(token)
+                    if idx == 0:
+                        word = word[idx + len(token) :]
+                        matched_tokens.append(j)
+                        a.append(token)
+                        last_matched = j
+                    j += 1
+
+                if not matched_tokens:
+                    # Reset j to most recent match
+                    j = last_matched
+                    index_scores[i] = 0.0
+                else:
+                    agg_grad = np.mean(gradient[matched_tokens], axis=0)
+                    index_scores[i] = np.linalg.norm(agg_grad, ord=1)
```

Comment:
So this is the sum of the absolute value of the gradients at each position? Is this how they do it in a paper? Just wondering. If so we might want to cite.

Reply:
I'm assuming you're talking about the HotFlip paper. But I thought HotFlip does not actually perform this type of greedy-word-wir type search? They do beam search. I don't recall any paper actually doing this. 

As for the L1 norm instead of L2 norm, I wasn't sure actually which one to pick and it wasn't clear from reading the papers. But I do know that [AllenInterpret](https://github.com/allenai/allennlp/blob/7c10fd26d0833e6f97f8559576ad759f4366a234/allennlp/interpret/saliency_interpreters/simple_gradient.py#L44) does L1 norm, so that's why I also chose it.